### PR TITLE
Sampled softmax

### PR DESCRIPTION
**Pull request by spirinamayya**
Added sampled softmax loss

Metrics
'model': 'sasrec_ssm_ids_only',
'HitRate@10': 0.30768741596715,
'MAP@10': 0.08537018545031581,
'AvgRecPopularity@10': 66279.45785366473,
'Serendipity@10': 3.32185414178862e-05,
'Intersection@10_popular': 0.5641220611214071,
'CoveredUsers@10': 1.0

### CODE REVIEW

File: rectools/models/nn/transformers/sasrec.py [hunk 1]
```diff
@@ -288,7 +288,7 @@ class SASRecModel(TransformerModelBase[SASRecModelConfig]):
     train_min_user_interactions : int, default 2
         Minimum number of interactions user should have to be used for training. Should be greater
         than 1.
-    loss : {"softmax", "BCE", "gBCE"}, default "softmax"
+    loss : {"softmax", "BCE", "gBCE", "sampled_softmax}, default "softmax"
```
Comment (blondered) [inline rectools/models/nn/transformers/sasrec.py:291]: Also in bert4rec

File: rectools/models/nn/transformers/lightning.py [hunk 2]
```diff
@@ -159,6 +159,11 @@ def training_step(self, batch: tp.Dict[str, torch.Tensor], batch_idx: int) -> to
             negatives = batch["negatives"]
             logits = self._get_pos_neg_logits(x, y, negatives)
             loss = self._calc_gbce_loss(logits, y, w, negatives)
+        elif self.loss == "sampled_softmax":
+            negatives = batch["negatives"]
+            logits = self._get_pos_neg_logits(x, y, negatives)
+            target = (y == 0).long()
```
Comment (blondered) [inline rectools/models/nn/transformers/lightning.py:165]: Why do we switch in this case? Do we really need it? Can we keep y=1 as real targets?

File: rectools/models/nn/transformers/sasrec.py (hunk 1 above)
Comment (spirinamayya) [reply rectools/models/nn/transformers/sasrec.py]: fixed

File: rectools/models/nn/transformers/lightning.py (hunk 2 above)
Comment (spirinamayya) [reply rectools/models/nn/transformers/lightning.py]: fixed

File: rectools/models/nn/transformers/lightning.py [hunk 3]
```diff
@@ -159,6 +159,12 @@ def training_step(self, batch: tp.Dict[str, torch.Tensor], batch_idx: int) -> to
             negatives = batch["negatives"]
             logits = self._get_pos_neg_logits(x, y, negatives)
             loss = self._calc_gbce_loss(logits, y, w, negatives)
+        elif self.loss == "sampled_softmax":
+            negatives = batch["negatives"]
+            logits = self._get_pos_neg_logits(x, y, negatives)
+            logits[:, :, [0, 1]] = logits[:, :, [1, 0]]
```
Comment (blondered) [inline rectools/models/nn/transformers/lightning.py:165]: Let's write a comment here `# We put positive logits at index 1 since index 0 is used to ignore padding`

File: rectools/models/nn/transformers/bert4rec.py [hunk 4]
```diff
@@ -208,7 +208,7 @@ class BERT4RecModel(TransformerModelBase[BERT4RecModelConfig]):
     train_min_user_interactions : int, default 2
         Minimum number of interactions user should have to be used for training. Should be greater
         than 1.
-    loss : {"softmax", "BCE", "gBCE"}, default "softmax"
+    loss : {"softmax", "BCE", "gBCE", "sampled_softmax"}, default "softmax"
```
Comment (blondered) [inline rectools/models/nn/transformers/bert4rec.py:211]: Please update changelog

File: rectools/models/nn/transformers/lightning.py (hunk 3 above)
Comment (spirinamayya) [reply rectools/models/nn/transformers/lightning.py:166]: fixed

File: rectools/models/nn/transformers/bert4rec.py (hunk 4 above)
Comment (spirinamayya) [reply rectools/models/nn/transformers/bert4rec.py:211]: fixed

File: CHANGELOG.md [hunk 5]
```diff
@@ -5,6 +5,10 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## Unreleased
+
+### Added
+- `sampled_softmax` loss [#274](https://github.com/MobileTeleSystems/RecTools/pull/274)
```
[the remark this answers is not in the document]
Comment (spirinamayya) [reply CHANGELOG.md]: fixed

**Review by blondered — APPROVED** [review]
